### ISSUE TITLE
Qualifier is sometimes not recognized

### ISSUE THREAD

**Issue by afdia** [mapstruct/mapstruct#767]
While working with qualifiers, I found a pretty bad bug, because it seems to choose the wrong mapping method when working with qualifiers.

I built a sample class with static inner classes to keep it concise, but the error also happens if the annotations and classes are in separate files (also String and Runnable are just some random classes to make the mapper require an explicit mapping method)
Example:
```
@Mapper
public abstract class ErrorMapper {
	@Qualifier @Target(ElementType.METHOD) @Retention(RetentionPolicy.SOURCE)
	public static @interface First {}

	@Qualifier @Target(ElementType.METHOD) @Retention(RetentionPolicy.SOURCE)
	public static @interface Second {}

	public static class A {
		private String a;
		public String getA() {return a;}
		public void setA(final String a) {this.a = a;}
	}
	public static class B {
		private Runnable b;
		public Runnable getB() {return b;}
		public void setB(final Runnable b) {this.b = b;}
	}
	@Mapping(source = "a", target = "b")
	public abstract B mapstructMapping(A in);

	@First
	protected Runnable manualMapping(final String in) {
		return null;
	}
}
```
This sample is working as expected (because I have not provided a mapping method without a qualifier for these 2 types) and mapstruct is telling me:
> Can't map property "java.lang.String a" to "java.lang.Runnable b". Consider to declare/implement a mapping method: "java.lang.Runnable map(java.lang.String value)".

If I change the @Mapping annotation to
```
@Mapping(source = "a", target = "b", qualifiedBy = First.class)
```
The manualMapping method is recognized and used as expected.

If I change the @Mapping annotation to
```
@Mapping(source = "a", target = "b", qualifiedBy = Second.class)
```
The manualMapping method is also used, although it is annotated with `@First` but not `@Second`
Therefore the mapping uses the wrong mapping method instead of printing an error that there is no no method for those 2 types which is annotated with `@Second`.

Furthermore if I remove the `@First` from the manualMapping  method, but let the mapping annotation unchanged as
```
@Mapping(source = "a", target = "b", qualifiedBy = Second.class)
```
it will also not print an error, although there is no mapping method which is qualified with `@Second`.
Although I could understand that unqualified methods work as fallback if there is no qualified method, I would prefer a strict handling of qualifications (which means if I request unqualified there must be an unqualified method and if I request a specific qualification, there must be a method with this qualification)

I hope the description was clear and there is no mistake or misunderstanding on my side about how qualifiers are expected to work.

**Comment by sjaakd** [comment]
@afdia Thanks for your feedback. The qualifier mechanism is conflict resolving mechanism. That is, if there's a conflict between potential candidates, the mechanism selects the most qualifying method. You can even add more qualifiers (the annotation allows for an array). See example 25 in the documentation, `@TitleTranslator`, `@EnglishToGerman` are both used to resolve the conflict.

In your case, if I understand you correctly: there's not conflict. So the qualifiers are not actually needed. 

However: possibly there could be a warning notice that you marked a class with a qualifier that is not actually used in the selection process.

**Comment by afdia** [comment]
Thanks for the quick response.

You are right, there is no conflict, but it's still somehow inconsistent, because in the first example I request a non-qualified conversion and it prints an error that there is none, which makes it look as if the qualifiers are strictly applied and not only for conflict resolution.

In general I think it would be better to treat mapping methods for qualified and unqualified types completely separate, because I cannot imagine a use case where I introduce a qualifier without actually wanting it to be used.
On the other hand the automatic fallback to the unqualified or differently qualified code can be dangerous if for some reason the qualified mapper-method (which could be located in a different mapper-class) is removed and instead of getting informed that it's missing now, it silently changes to another mapping.
This was in fact what happened to me ;)

It would probably also be more in line with the usage of JSR330 (dependency injection) qualifiers.
I know that they live in a different namespaces but #750 introduces a `@Named` annotation too which may suggest to the user that you use them in a similar way and as far as I know these annotations are always applied.

After reading some more issues, one of the error cases has already been reported in #533, so this issue is partially a duplicate, but perhaps the example is still helpful.

**Comment by sjaakd** [comment]
You are right our naming is inspired by JSR330 and hence would imply similar behavior by users familiar with JSR330. 

However, qualifiers are not the only mechanism used in selection. There are a few (source type, target type,  result-type, qualifier, jaxb naming) , each of them applied subsequently. I'm not to familiar with JSR330, but I'm not sure whether these kind of aspects also apply to JSR330 and whether the similarity holds entirely. Perhaps some documentation warning somewhere would help here.

When looking at a warning of non-used selection (explicit) selection criteria, we could emit a warning. In the case of qualifiers, that could actually go into two directions: a to-be-generated-mapping-method is referring to a qualifier that is not being used. And, there are qualifiers on the class-path that are never actually referred to. The first one is I guess the most pressing and easy to resolve. The second one is more tricky.

Next to that, we should perhaps also emit a warning for other explicit non used selection criteria.

**Comment by afdia** [comment]
Sounds good.
I've updated the issue title and posted a short intro to make clear that my description as a bug was probably not correct.

**Comment by maposchmi** [comment]
Could it be potential "dangerous" that you have that dead code, as soon as the mapstruct-implementation changes its behavior regarding to that qualifiers?
